packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -105,6 +115,130 @@ export class ExecutionEngine {
     };
   }
 
+  private async _checkNonces(state: ExecutionEngineState, futures: Future[]) {
+    const transactionSubmittingFutures = futures.filter(
+      isFutureThatSubmitsOnchainTransaction
+    );
+
+    const usedAccounts = uniq(
+      transactionSubmittingFutures.map((f) => resolveFromAddress(f.from, state))
+    );
+
+    const pendingIgnitionTransactions = transactionSubmittingFutures
+      .map((f) => {
+        const exState = state.executionStateMap[f.id];
+
+        if (
+          exState === undefined ||
+          exState.onchain.currentExecution === null ||
+          exState.onchain.nonce === null ||
+          exState.onchain.txHash === null
+        ) {
+          return null;
+        }
+
+        return {
+          futureId: f.id,
+          executionId: exState.onchain.currentExecution,
+          from: resolveFromAddress(f.from, state),
+          nonce: exState.onchain.nonce,
+          hash: exState.onchain.txHash,
+        };
+      })
+      .filter(<T>(x: T | null): x is T => x !== null);
+
+    const inflightByAccount = groupBy(pendingIgnitionTransactions, "from");
+
+    for (const usedAccount of usedAccounts) {
+      const pending = await state.chainDispatcher.getPendingTransactionCount(
+        usedAccount
+      );
+      const latest = await state.chainDispatcher.getLatestTransactionCount(
+        usedAccount
+      );
+
+      const inflight = inflightByAccount[usedAccount];
+
+      // no inflight transactions
+      if (inflight === undefined) {
+        if (pending !== latest) {
+          throw new IgnitionError(
+            `Pending transactions for account: ${usedAccount}, please wait for transactions to complete before running a deploy`
+          );
+        }
+
+        // no pending transactions
+        return;
+      }
+
+      const maxNonceTran = maxBy(inflight, "nonce");
+
+      assertIgnitionInvariant(
+        maxNonceTran !== undefined,
+        "Always at least one entry going into maxBy"
+      );
+
+      // The pending transactions do not match the expected list
+      // given the inflight transactions from the previous run
+      if (maxNonceTran.nonce + 1 < pending) {
```

Comment:
I don't understand this condition. Can we name them? maybe

`const morePendingTransactionsThanWeExpected = maxNonceTran.nonce < pending`

Is that what we are trying to check?

I think we can do better in this case, but let's leave it as it is for now.

Reply:
I have pulled out the condition as a private function so we can name it: `_isMorePendingThanPreviousRunIndicates`